### ISSUE TITLE
What Does Haskell Have to Do with C?

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2938]
What Does Haskell Have to Do with C++?<br>
<br>
https://ift.tt/2foAPXX<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<p>If you want to understand C++ template metaprogramming (TMP) you have to know functional programming. Seriously. I want you to think of TMP as maximally obfuscated (subset of) Haskell, and I&rsquo;ll illustrate this point by point. If you don&rsquo;t know Haskell, don&rsquo;t worry, I&rsquo;ll explain the syntax as I go.</p>
<p>The nice thing about single-paradigm languages like Haskell is that they have very simple syntax (think of Lisp that does everything with just a bunch of parentheses). I will start the Haskell-C++TMP mapping with basics, like functions and recursion, but I&rsquo;ll try to cover a lot more, including higher-order functions, pattern matching, list comprehension (did you know it was expressible in C++?), and more.</p>
<p>Keep in mind that my Haskell examples are runtime functions operating on runtime data whereas their C++ TMP equivalents are compile-time templates operating mostly on types. Operation on types are essential in providing correct and efficient implementations of parametrized classes and functions.</p>
<p>By necessity the examples are simple, but the same mapping may be applied to much more complex templates from the C++ Standard Library and Boost. As a bonus, I&rsquo;ll also explain the hot new thing, variadic templates.</p>
<h2>Functional Approach to Functions</h2>
<p>How do you implement useful functions if you don&rsquo;t have mutable variables, <var>if</var> statements, or loops? To a C++ programmer that might seem like an impossible task. But that&rsquo;s the reality of C++ compile-time language that forms the basis of TMP. Functional programming to the rescue!</p>
<p>As a warm-up, let&rsquo;s see how Haskell implements a simple function, the factorial:</p>
<pre>
fact 0 = 1
fact n = n * fact (n - 1)
</pre>
<p>The first line states that the factorial of zero is one. The second line defines factorial for a non-zero argument, n (strictly speaking it only works for positive non-zero arguments). It does it using recursion: factorial of <var>n</var> is equal to <var>n</var> times the factorial of <var>n-1</var>. The recursion stops when <var>n</var> is equal to zero, in which case the first definition kicks in. Notice that the definition of the function <var>fact</var> is split into two sub-definitions. So when you call:</p>
<pre>
fact 4
</pre>
<p>the first definition is looked up first and, if it doesn&rsquo;t match the argument (which it doesn&rsquo;t), the second one comes into play. This is the simplest case of <i>pattern matching</i>: 4 doesn&rsquo;t match the &ldquo;pattern&rdquo; 0, but it matches the pattern <var>n</var>.</p>
<p>Here&rsquo;s almost <i>exactly</i> the same code expressed in C++ TMP:</p>
<pre>
template<span>&lt;int n&gt;</span> struct
<span>fact</span> {
    static const int value = <span>n * fact&lt;n - 1&gt;</span>::value;
};

template&lt;&gt; struct
<span>fact&lt;0&gt;</span> { // specialization for n = 0
    static const int value = <span>1</span>;
};
</pre>
<p>You might notice how the horrible syntax of C++ TMP obscures the simplicity and elegance of this code. But once you are equipped with the C++/Haskell decoder ring, things become a lot clearer.</p>
<p>Let&rsquo;s analyze this code. Just like in Haskell, there are two definitions of <var>fact</var>, except that their order is inverted. This is because C++ requires template specialization to follow the template&rsquo;s general definition (or declaration, as we&rsquo;ll see later). The pattern matching of arguments in C++ does not follow the order of declarations but rather is based on &ldquo;best match&rdquo;. If you instantiate the template with argument zero:</p>
<pre>
cout &lt;&lt; "Factorial of 0 = " &lt;&lt; fact&lt;0&gt;::value &lt;&lt; endl;
</pre>
<p>the second pattern, <var>fact&lt;0&gt;</var>, is a better fit. Otherwise the first one, <var>&lt;int n&gt;</var>, is used.</p>
<p>Notice also the weird syntax for &ldquo;function call&rdquo;</p>
<pre>
fact&lt;n&gt;::value
</pre>
<p>and for the &ldquo;return statement&rdquo;</p>
<pre>
static const int value = n * fact&lt;n - 1&gt;::value;
</pre>
<p>This all makes sense if you look at templates as definitions of parameterized types, which was their initial purpose in C++. In that interpretation, we are defining a <var>struct</var> called <var>fact</var>, parameterized by an integer <var>n</var>, whose sole member is a static const integer called <var>value</var>. Moreover, this template is specialized for the case of <var>n</var> equal zero.</p>
<p>Now I want you to forget about what I just said and put on the glasses which make the C++ code look like the corresponding Haskell code.</p>
<p>Here&rsquo;s another example&ndash;this time of a predicate (a function returning a Boolean):</p>
<pre>
is_zero 0 = True
is_zero x = False
</pre>
<p>Let&rsquo;s spice it up a little for C++ and define a predicate on types rather than integers. The following compile-time function returns <var>true</var> only when the type <var>T</var> is a pointer:</p>
<pre>
template&lt;class T&gt; struct
<span>isPtr</span> {
    static const bool value = false;
};

template&lt;class U&gt; struct
<span>isPtr&lt;U*&gt;</span> {
    static const bool value = true;
};
</pre>
<p>This time the actual argument to <var>isPtr</var> is first matched to the more specialized pattern, <var>U*</var> and, if it fails, the general pattern is used.</p>
<p>We can add yet another specialization, which will pattern-match a const pointer:</p>
<pre>
template&lt;class U&gt; struct
<span>isPtr&lt;U * const&gt;</span> {
    static const bool value = true;
};
</pre>
<p>These types of type predicates may be, for instance, used to select more flexible and efficient implementations of parameterized containers. Think of the differences between a vector of values vs. a vector of pointers.</p>
<h2>Lists</h2>
<p>The basic data structure in functional languages is the list. Haskell&rsquo;s lists are introduced using square brackets. For instance, a list of three numbers, 1, 2, 3, looks like:</p>
<pre>
[1, 2, 3]
</pre>
<p>List processing in functional languages follows the standard pattern: a list is split into <var>head</var> and <var>tail</var>, an operation is performed on the head, and the tail is processed using recursion. The splitting is done by pattern matching: the Haskell pattern being <var>(head:tail)</var> (to be precise, the colon in parentheses represents the <var>cons</var> operation&ndash;the creation of a list by prepending an element to an existing list).</p>
<p>Here&rsquo;s a simple function, <var>count</var>, that calculates the length of a list:</p>
<pre>
count [] = 0
count (head:tail) = 1 + count tail
</pre>
<p>The first pattern, [], matches an empty list; the second a non-empty one. Notice that a function call in Haskell doesn&rsquo;t use parentheses around arguments, so <var>count tail</var> is interpreted as a call to <var>count</var> with the argument <var>tail</var>.</p>
<p>Before C++0x, TMP was severely crippled by the lack of a list primitive. People used separate definitions for a list of zero, one, two, etc., elements, and even used special macros to define them. This is no longer true in C++0x, thanks to <i>variadic templates</i> and <i>template parameter packs</i>. Here&rsquo;s our Haskel <var>count</var> translated into C++ TMP:</p>
<pre>
// Just a declaration
template&lt;class... list&gt; struct
<span>count</span>;

template&lt;&gt; struct
<span>count&lt;&gt;</span> {
    static const int value = 0;
};

template&lt;class head, class... tail&gt; struct
<span>count&lt;head, tail...&gt;</span> {
    static const int value = 1 + count&lt;tail...&gt;::value;
};
</pre>
<p>First we have a declaration (not a definition) of the template <var>count</var> that takes a variable number of type parameters (the keyword <var>class</var> or <var>typename</var> introduces a type parameter). They are packed into a template parameter pack, <var>list</var>.</p>
<p>Once this general declaration is visible, specializations may follow in any order. I arranged them to follow the Haskell example. The first one matches the empty list and returns zero. The second one uses the pattern, <var>&lt;head, tail&hellip;&gt;</var>. This pattern will match any non-empty list and split it into the <var>head</var> and the (possibly empty) <var>tail</var>.</p>
<p>To &ldquo;call&rdquo; a variadic template, you initiate it with an arbitrary number of arguments and retrieve its member, <var>value</var>, e.g.,</p>
<pre>
int n = count&lt;int, char, long&gt;::value; // returns 3
</pre>
<p>A few words about variadic templates: A variadic template introduces a template parameter pack using the notation <var>class&hellip; pack</var> (or <var>int&hellip; ipack</var>, etc&hellip;). The only thing you may do with a pack is to expand it and pass to another variadic template. The expansion is done by following the name of the pack with three dots, as in <var>tail&hellip;</var>. You&rsquo;ll see more examples later.</p>
<p>Variadic templates have many applications such as type-safe printf, tuples (objects that store an arbitrary number of differently typed arguments), variants, and many more.</p>
<h2>Higher-Order Functions and Closures</h2>
<p>The real power of functional programming comes from treating functions as first class citizens. It means that you may pass functions to other functions and return functions from functions. Functions operating on functions are called higher-order functions. Surprisingly, it seems like compile-time C++ has better support for higher-order functions than run-time C++.</p>
<p>Let&rsquo;s start with a Haskell example. I want to define a function that takes two predicate functions and returns another predicate function that combines the two using logical <var>OR</var>. Here it is in Haskell:</p>
<pre>
or_combinator f1 f2 = 
    &lambda; x -&gt; (f1 x) || (f2 x)
</pre>
<p>The <var>or_combinator</var> returns an anonymous function (the famous &ldquo;lambda&rdquo;) that takes one argument, <var>x</var>, calls both <var>f1</var> and <var>f2</var> with it, and returns the logical <var>OR</var> of the two results. The return value of <var>or_combinator</var> is this freshly constructed function. I can then call this function with an arbitrary argument. For instance, here I&rsquo;m checking if 2 is either zero or one (guess what, it isn&rsquo;t!):</p>
<pre>
(or_combinator is_zero is_one) 2
</pre>
<p>I put the parentheses around the function and its arguments for readability, although they are not strictly necessary. 2 is the argument to the function returned by <var>or_combinator</var>.</p>
<p>The lambda that&rsquo;s returned from <var>or_combinator</var> is actually a closure. It &ldquo;captures&rdquo; the two arguments, <var>f1</var> and <var>f2</var> passed to <var>or_combinator</var>. They may be used long after the call to <var>or_combinator</var> has returned.</p>
<p>It might take some getting used to it before you are comfortable with functions taking functions and returning functions, but it&rsquo;s much easier to learn this stuff in Haskell than in the obfuscated C++. Indeed, here&rsquo;s an almost direct translation of this example:</p>
<pre>
template&lt;template&lt;class&gt; class <span>f1</span>, template&lt;class&gt; class <span>f2</span>&gt; struct
<span>or_combinator</span> {
    template&lt;class T&gt; struct
    <span>lambda</span> {
        static const bool value = <span>f1&lt;T&gt;</span>::value <span>|| f2&lt;T&gt;</span>::value;
    };
};
</pre>
<p>Since in the metalanguage a function is represented by a template, the template <var>or_combinator</var> takes two such templates as arguments. It &ldquo;calls&rdquo; these templates using the standard syntax <var>f&lt;T&gt;::value</var>. Actually, the <var>or_combinator</var> doesn&rsquo;t call these functions. Instead it defines a new template, which I call <var>lambda</var>, that takes the argument <var>T</var> and calls those functions. This template acts like a closure&ndash;it captures the two templates that are the arguments to <var>or_combinator</var>.</p>
<p>Here&rsquo;s how you may use the <var>or_combinator</var> to combine two tests, <var>isPtr</var> and <var>isConst</var> and apply the result to the type <var>const int</var>:</p>
<pre>
std::cout 
   &lt;&lt; "or_combinator&lt;isPtr, isConst&gt;::lambda&lt;const int&gt;::value = "
   &lt;&lt; or_combinator&lt;isPtr, isConst&gt;::lambda&lt;const int&gt;::value 
   &lt;&lt; std::endl;
</pre>
<p>Such logical combinators are essential for predicate composability.</p>
<h2>Higher-Order Functions Operating on Lists</h2>
<p>Once you combine higher-order functions with lists you have a powerful functional language at your disposal. Higher-order functions operating on lists look very much like algorithms. Let me show you some classic examples. Here&rsquo;s the function (or algorithm), <var>all</var>, that returns <var>true</var> if and only if all elements of a list satisfy a given predicate.</p>
<pre>
all pred [] = True
all pred (head:tail) = (pred head) &amp;&amp; (all pred tail)
</pre>
<p>By now you should be familiar with all the techniques I used here, like pattern matching or list recursion.</p>
<p>Here&rsquo;s the same code obfuscated by the C++ syntax:</p>
<pre>
template&lt;template&lt;class&gt; class predicate, class... list&gt; struct
<span>all</span>;

template&lt;template&lt;class&gt; class predicate&gt; struct
<span>all&lt;predicate&gt;</span> {
    static const bool value = true;
};

template&lt;
    template&lt;class&gt; class predicate, 
    class head, 
    class... tail&gt; struct
<span>all&lt;predicate, head, tail...&gt;</span> {
    static const bool value = 
        predicate&lt;head&gt;::value 
        &amp;&amp; all&lt;predicate, tail...&gt;::value;
};
</pre>
<p>Except for the initial declaration required by C++ there is a one-to-one paradigm match between the two implementations.</p>
<p>Another useful algorithm, a veritable workhorse of functional programming, is &ldquo;fold right&rdquo; (together with it&rsquo;s dual partner, &ldquo;fold left&rdquo;). It folds a list while accumulating the results (that&rsquo;s why in runtime C++ this algorithm is called &ldquo;accumulate&rdquo;). Here&rsquo;s the Haskell implementation:</p>
<pre>
foldr f init [] = init
foldr f init (head:tail) =
    f head (foldr f init tail)
</pre>
<p>Function <var>f</var>, which is the first argument to <var>foldr</var>, takes two arguments, the current element of the list and the accumulated value. Its purpose is to process the element and incorporate the result in the accumulator. The new accumulated value is then returned. It is totally up to the client to decide what kind of processing to perform, how it is accumulated, and what kind of value is used. The second argument, <var>init</var>, is the initial value for the accumulator.</p>
<p>Here&rsquo;s how it works: The result of <var>foldr</var> is generated by acting with <var>f</var> on the head of the list and whatever has been accumulated by processing the tail of the list. The algorithm recurses until the tail is empty, in which case it returns the initial value. At runtime this type of algorithm would make N recursive calls before starting to pop the stack and accumulate the results.</p>
<p>For instance, <var>foldr</var> may be used to sum the elements of a list (<var>so_far</var> is the accumulator, which is initialized to zero):</p>
<pre>
add_it elem so_far = elem + so_far
sum_it lst = foldr add_it 0 lst
</pre>
<p>The accumulator function is <var>add_it</var>. If, instead, I wanted to calculate the product of all elements, I&rsquo;d use a function <var>mult_it</var> and the starting value of one. You get the idea.</p>
<p>Here&rsquo;s the same algorithm in C++ TMP:</p>
<pre>
template&lt;template&lt;class, int&gt; class, int, class...&gt; struct
<span>fold_right</span>;

template&lt;template&lt;class, int&gt; class f, int init&gt; struct
<span>fold_right&lt;f, init&gt;</span> {
    static const int value = init;
};

template&lt;template&lt;class, int&gt; class f, int init, class head, class...tail&gt; struct
<span>fold_right&lt;f, init, head, tail...&gt;</span> {
    static const int value = f&lt;head, fold_right&lt;f, init, tail...&gt;::value&gt;::value;
};
</pre>
<p>Once you understand the Haskell version, this complex code suddenly becomes transparent (if it doesn&rsquo;t, try squinting <img alt="&#128521;" src="https://s0.wp.com/wp-content/mu-plugins/wpcom-smileys/twemoji/2/svg/1f609.svg"> ).</p>
<h2>Lists of Numbers</h2>
<p>Let&rsquo;s now switch to integers for a moment. Haskell defines a function <var>sum</var> that adds all elements of a list:</p>
<pre>
sum [] = 0
sum (head:tail) = head + (sum tail)
</pre>
<p>We can do the same in C++ TMP (in five times as many lines of code):</p>
<pre>
template&lt;int...&gt; struct
<span>sum</span>;

template&lt;&gt; struct
<span>sum&lt;&gt;</span> {
    static const int value = 0;
};

template&lt;int i, int... tail&gt; struct
<span>sum&lt;i, tail...&gt;</span> {
    static const int value = i + sum&lt;tail...&gt;::value;
};
</pre>
<h2>List Comprehension</h2>
<p>Haskell has one more trick up its sleeve for operating on lists without explicit recursion. It&rsquo;s called <i>list comprehension</i>. It&rsquo;s a way of defining new lists based on existing lists. The nomenclature and notation are borrowed from Set Theory, where you often encounter definitions such as: S is a set of elements, where&hellip; Let&rsquo;s look at a simple Haskell example:</p>
<pre>
[x * x | x &lt;- [3, 4, 5]]
</pre>
<p>This is a set (list) of elements <var>x * x</var>, where <var>x</var> is from the list <var>[3, 4, 5]</var>.</p>
<p>Remember our recursive definition of <var>count</var>? Using list comprehension it&rsquo;s reduced to a one-liner:</p>
<pre>
count lst = sum [1 | x &lt;- lst]
</pre>
<p>Here, we create a list of ones, one for each element of the list. Our result is the sum of those ones. To make this definition more amenable to translation into C++, let&rsquo;s define an auxiliary function <var>one</var> that, for any argument <var>x</var>, returns 1.</p>
<pre>
one x = 1
</pre>
<p>Here&rsquo;s the modified definition of <var>count</var>:</p>
<pre>
count lst = sum [one x | x &lt;- lst]
</pre>
<p>Now we are ready to convert this code to C++ TMP:</p>
<pre>
template&lt;class T&gt; struct
<span>one</span> {
    static const int value = 1;
};

template&lt;class... lst&gt; struct
<span>count</span> {
    static const int value = sum&lt;one&lt;lst&gt;::value...&gt;::value;
};
</pre>
<p>Here our list is stored in a template parameter pack, <var>lst</var>. If we wanted to expand this pack, we&rsquo;d use the notation <var>lst&hellip;</var>, but that&rsquo;s not what&rsquo;s happening here. The ellipsis appears after the pattern containing the pack:</p>
<pre>
one&lt;lst&gt;::value...
</pre>
<p>Compare this with the equivalent Haskell:</p>
<pre>
[one x | x &lt;- lst]
</pre>
<p>In C++, when the ellipsis follows a pattern that contains a pack, it&rsquo;s not the pack that&rsquo;s expanded, but the whole pattern is repeated for each element of the pack. Here, if our list were <var>&lt;int, char, void*&gt;</var>, the pattern would be expanded to:</p>
<pre>
&lt;one&lt;int&gt;::value, one&lt;char&gt;::value, one&lt;void*&gt;::value&gt;
</pre>
<p>The subsequent call to <var>sum</var> would be made with those arguments.</p>
<p>Notice that a different positioning of the ellipsis would result in a completely different expansion. This pattern:</p>
<pre>
one&lt;lst...&gt;::value
</pre>
<p>would result in the call to <var>one</var> with the list of types, which would be an error.</p>
<p>Here&rsquo;s another example of pattern expansion: a function that counts the number of pointers in a list of types:</p>
<pre>
template&lt;class... lst&gt; struct
<span>countPtrs</span> {
    static const int value = sum&lt;isPtr&lt;lst&gt;::value ...&gt;::value;
};
</pre>
<p>In this case the pattern is:</p>
<pre>
isPtr&lt;lst&gt;::value ...
</pre>
<p>and it expands into a list of Booleans. (I&rsquo;m taking advantage of the fact that <var>false</var> is zero and <var>true</var> is one, when converted to integers.)</p>
<p>You may find a more complex practical example in the Gregor, J&auml;rvi, and Powell paper (see bibliography).</p>
<h2>Continuations</h2>
<p>List comprehension can be used to define some very useful higher-order functions. One of such functions is <var>map</var>, which takes a list and applies a unary function to each element, resulting in a new list. You might be familiar with the runtime implementation of this algorithm in C++ STL under the name of <var>transform</var>. This is what <var>map</var> looks like in Haskell:</p>
<pre>
map f lst = [f x | x &lt;- lst]
</pre>
<p>Here, <var>f</var> is the unary function and <var>lst</var> is the input list. You have to admire the terseness and elegance of this notation.</p>
<p>The first impulse would be to translate it into C++ TMP as:</p>
<pre>
template&lt;template&lt;class&gt; class f, class... lst&gt; struct
<span>map</span> {
    typedef f&lt;lst&gt;... type;
};
</pre>
<p>This is surprisingly terse too. The problem is that it doesn&rsquo;t compile. As far as I know there is no way for a template to &ldquo;return&rdquo; a variable list of elements. In my opinion, this is a major language design flaw, but that&rsquo;s just me.</p>
<p>There are several workarounds, none of them too exciting. One is to define a separate entity called a <var>typelist</var> along the lines of:</p>
<pre>
template struct
<span>typelist&lt;hd, tl...&gt;</span> {
    typedef hd head;
    typedef typelist&lt;tl...&gt; tail;
};
</pre>
<p>(As a matter of fact I have implemented typelists and related algorithms both in C++ and D.)</p>
<p>Another approach is to use continuations. Template parameter packs cannot be returned, but they can be passed to variadic templates (after expansion). So how about defining an algorithm like <var>map</var> to take one additional function that would consume the list that is the result of mapping? Such a function is often called a continuation, since it continues the calculation where normally one would return the result. First, let&rsquo;s do it in Haskell:</p>
<pre>
map_cont cont f lst = cont [f x | x &lt;- lst]
</pre>
<p>The function <var>map_cont</var> is just like <var>map</var> except that it takes a continuation, <var>cont</var>, and applies it to the result of mapping. We can test it by defining yet another implementation of <var>count</var>:</p>
<pre>
count_cont lst = map_cont sum one lst
</pre>
<p>The continuation here is the function <var>sum</var> that will be applied to the list produced by acting with function <var>one</var> on the list <var>lst</var>. Since this is quite a handful, let me rewrite it in a more familiar notation of runtime C++:</p>
<pre>
int map_cont(int (*cont)(list), int (*f)(int), list lst) {
    list tmp;
    for (auto it = lst.begin(); it != lst.end(); ++it)
        tmp.push_front(f(*it));
    return cont(tmp);
}
</pre>
<p>Now for the same thing in compile-time C++:</p>
<pre>
template&lt;template&lt;class...&gt; class cont, 
         template&lt;class&gt; class f,
         class... lst&gt; struct
<span>map_cont</span> {
    static const int value =
        cont&lt;typename f&lt;lst&gt;::type ...&gt;::value;
};
</pre>
<p>It&rsquo;s a one-to-one mapping of Haskell code&ndash;and it has very little in common with the iterative runtime C++ implementation. Also notice how loose the typing is in the TMP version as compared with the runtime version. The continuation is declared as a variadic template taking types, function <var>f</var> is declared as taking a type, and the list is a variadic list of types. Nothing is said about return types, except for the constraints that <var>f</var> returns a type (because <var>cont</var> consumes a list of types) and <var>cont</var> returns an <var>int</var>. Actually, this last constraint can be relaxed if we turn integers into types&ndash;a standard trick (hack?) in TMP:</p>
<pre>
template&lt;int n&gt; struct
<span>Int</span> {
    static const int value = n;
};
</pre>
<p>Loose typing&ndash;or &ldquo;kinding,&rdquo; as it is called for types of types&ndash;is an essential part of compile-time programming. In fact the popularity of the above trick shows that C++ kinding might be already too strong.</p>
<h2>The D Digression</h2>
<p>I&rsquo;m grateful to Andrei Alexandrescu for reviewing this post. Since he objected to the sentence, &ldquo;I&rsquo;m disappointed that the D programming language followed the same path as C++ rather than lead the way,&rdquo; I feel compelled to support my view with at least one example. Consider various implementations of <var>all</var>.</p>
<p>In Haskell, beside the terse and elegant version I showed before:</p>
<pre>
all pred [] = True
all pred (head:tail) = (pred head) &amp;&amp; (all pred tail)
</pre>
<p>there is also a slightly more verbose one:</p>
<pre>
all pred list =
    if null list then True
    else pred (head list) &amp;&amp; all pred (tail list)
</pre>
<p>which translates better into D. The D version (taken from its standard library, Phobos) is not as short as Haskell&rsquo;s, but follows the same functional paradigm:</p>
<pre>
template allSatisfy(alias F, T...) {
    static if (T.length == 1)
    {
        alias F!(T[0]) <span>allSatisfy</span>;
    }
    else
    {
        enum bool <span>allSatisfy</span> = F!(T[0]) &amp;&amp; allSatisfy!(F, T[1 .. $]);
    }
}
</pre>
<p>It definitely beats C++, there&rsquo;s no doubt about it. There are some oddities about it though. Notice that the type tuple, <var>T&hellip;</var> gets the standard list treatment, but the split into the head and tail follows the array/slice notation. The head is <var>T[0]</var> and the tail is an array slice, <var>T[1..$]</var>. Instead of using <var>value</var> for the return value, D uses the &ldquo;eponymous hack,&rdquo; as I call it. The template &ldquo;returns&rdquo; the value using its own name. It&rsquo;s a hack because it breaks down if you want to define the equivalent of &ldquo;local variable&rdquo; inside a template. For instance, the following code doesn&rsquo;t compile:</p>
<pre>
template allSatisfy(alias F, T...) {
    static if (T.length == 1)
    {
        alias F!(T[0]) allSatisfy;
    }
    else
    {
        <span>private enum bool tailResult = allSatisfy!(F, T[1..$]);</span>
        enum bool allSatisfy = F!(T[0]) &amp;&amp; tailResult;
    }
}
</pre>
<p>This breaks one of the fundamental property of any language: decomposability. You want to be able to decompose your calculation into smaller chunks and then combine them together. Of course, you may still use decomposition if you don&rsquo;t use the eponymous hack and just call your return value <var>value</var>. But that means modifying all the calling sites.</p>
<p>By the way, this is how decomposition works in Haskell:</p>
<pre>
all2 pred [] = True;
all2 pred (head:tail) = (pred head) &amp;&amp; tailResult
    where tailResult = all2 pred tail
</pre>
<p>Anyway, the important point is that there is <i>no reason</i> to force functional programming paradigm at compile-time. The following hypothetical syntax would be much easier for programmers who are used to imperative programming:</p>
<pre>
template allSatisfy(alias Pred, T...) {
    foreach(t; T)
        if (!Pred!(t))
            return false;
    return true;
}
</pre>
<p>In fact it&rsquo;s much closer to the parameterized runtime function proposed by Andrei:</p>
<pre>
bool all(alias pred, Range)(Range r) {
    foreach (e; r) 
        if (!pred(e)) 
            return false;
    return true;
}
</pre>
<p>This is why I&rsquo;m disappointed that the D programming language followed the same path as C++ rather than lead the way.</p>
<h2>Conclusions</h2>
<p>I have argued that some familiarity with Haskell may be really helpful in understanding and designing templates in C++. You might ask why C++ chose such horrible syntax to do compile-time functional programming. Well, it didn&rsquo;t. The ability to do compile-time calculations in C++ was <i>discovered</i> rather than built into the language. It was a very fruitful discovery, as the subsequent developments, especially the implementation of the Boost MPL, have shown. However, once the functional paradigm and its weird syntax took root in C++ TMP, it stayed there forever. I&rsquo;m aware of only one effort to rationalize C++ TMP by Daveed Vandevoorde in <a href="http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2003/n1471.pdf">Reflective Metaprogramming in C++</a>.</p>
<p>I have tested all code in this blog using the Hugs interpreter for Haskell and the GNU C++ compiler v. 4.4.1 with the special switch <var>-std=c++0x</var>. The names I used in the blog might conflict with the standard definitions. For instance, Hugs defines its own <var>map</var> and <var>foldr</var>.</p>
<p>If you want to learn more about template metaprogramming, I recommend two books:</p>
<ol>
<li>Andrei Alexandrescu, Modern C++ Design</li>
<li>David Abrahams and Aleksey Gurtvoy, C++ Template Metaprogramming</li>
</ol>
<p>The reference I used for variadic templates was the paper by <a href="http://www.osl.iu.edu/~dgregor/cpp/variadic-templates.pdf">Douglas Gregor, Jaakko J&auml;rvi, and Gary Powell</a></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via &nbsp;&nbsp;Bartosz Milewski's Programming Cafe https://ift.tt/2a8JGGc<br>
<br>
March 12, 2019 at 02:36PM
